# Week E: Student-led experiments

In week E students can organise their own experiments to conclude the research project. Many students decide to return to the lab and test their predictions of antibiotic resistance by treating the bacteria with antibiotics. Students are encouraged to work in small groups (2-4) rather than individually, and these are most likely the presentation groups from week D.

## Experimental Plan

Students must submit a written time plan with what they intend to do in each half day of week E (or a gantt chart). They must also prepare a detailed list of materials (media, antibiotics, equipment etc) 5 weeks before week E to ensure ordering and preparation can be completed.

The experiment frequently consists of students making plates with Mueller Hinton broth, starting a liquid culture of the test strains, spreading plates with test strains, placing commercial antibiotic test disks onto the spread plate, measuring the zone of growth inhibition around the disk.

This plan is discussed with staff to ensure feasibility. We encourage simple experiments that can be completed to a high standard over complex experiments that may not successful in the limited time available.

There is a page of more student-facing instructions in [experimental_plan.md](experimental_plan.md)

## Manuscript Help

In week E we start to coach students on the best way to write a scientific manuscript for their final assessment.

## Employability Skills

In week E, in addition to help with the manuscript plan, there will be time with students to translate their course experiences into named employability skills for a CV. These may include:

- Microbiology techniques. Plating, culturing and gram staining bacteria from clinical isolates, antibiotic resistance tests

- Bacterial DNA extraction and quantification. Construction of DNA sequencing library, loading and running of an ONT minION nanopore DNA sequencer

- Knowledge of the unix command line. Basic command line skills to navigate, inspect and manipulate large data sets. Use of the command line for bioinformatics

- Genome assembly and quality control. Ability to detect contamination in sequence sets. Basic skills to assemble a bacterial genome from DNA sequence data, and assess the quality of the assembly

- Genome annotation and AMR gene detection.  Basic skills to annotate a bacterial genome and to predict antimicrobial resistance genes
