Understanding how to do well with the course

Welcome to BGS. Each lab week will have a separate lab manual with all
the details and instructions you need. This is a guide to the BGS course
structure, and a way in which we can highlight the important stuff,
guiding you (we hope) to do well.

601137 Bacterial Genome Sequencing is a 40-credit 2-semester course and
part of the final-year-projects umbrella of modules. We will work
together to characterise bacterial species from a clinical environment.
We will prepare DNA, construct a DNA sequencing library, and run a
modern minION DNA sequencer in the lab to generate DNA sequences and
construct a genome assembly. In semester 2 we will annotate the
bacterial genome to identify genes involved in antibiotic resistance.
Finally you can test your personal assessment of the bacterial
resistance to specific antibiotics (determined from the sequence) by
applying antibiotics. You will have lots to write up.

There will be **five full weeks in the lab** spread over 2 semesters.
**This is not a practical class**, this is real scientific research. You
are doing things that have never been done before, analysing totally new
data, and drawing unique conclusions. Scientists do not work in
isolation, you will be working in groups, working with academic staff,
helping out others, and learning as a team. We have designed this so
that your fate is in your own hands though, your assessments are mostly
individual and you will not be dependent on others for good marks.

## Goals of the Course

Try to think what you want to get out of this course. Perhaps you just
want 40 credits, perhaps to learn very modern lab skills, perhaps to
sequence a bacterial genome because it\'s fun and fascinating. There are
lots of other things you might want.

We want to teach a good course. We want to provide an alternative to the
standard IRP project format to promote a diversity of learning methods.
We want to interact with students and guide them through research
because it\'s fun and rewarding and something we don't often get to do.
We want to teach students cutting edge lab techniques to prepare them
for research or industry. We want to sequence bacterial genomes because
it\'s fun and fascinating.

We can arrange this course so that both students and staff get what they
want, but we need to collaborate. We are working hard in the background
to prepare everything and to teach the material. You will also need to
work hard behind the scenes reading papers, learning methods and skills,
thinking and asking questions. You can get great marks on this course
(and remember it\'s 40 credits) but it will require very strong
engagement, you will have to work for it.

## Assessments

1.  Portfolio 25%

2.  Group presentation 25%

3.  Manuscript 50%

The assessments are *briefly* outlined below but we will explain and
guide you much more as the course progresses. Each assessment will have
a help page on Canvas and also a help session before it is due to be
submitted.

### Portfolio

You will individually prepare and submit a document (portfolio)
outlining the different work you have done. This has two compulsory
components

1.  **Electronic Lab Notebook**, a reproducible record of all your
    > research activities

2.  **Evidence of achievement of competencies**, briefly document the
    > completion of specific research activities that are assigned to
    > you (eg run a gel, summarise a paper, trim DNA sequences)

### Group presentation

You will present your work as a small group (\~4 students) in a 10
minute talk to the other students. You will present your preliminary
analyses and results describing what they mean. You are not assessed on
how slick you are at presenting, but more on how well you explain the
system and results to the class. This assessment is a great way to get
detailed feedback that will feed into your final assessment, the
scientific manuscript. We have run these for years and it is both
teaching and assessment, you will learn a lot during the discussions.
Also, there may be biscuits.

### Manuscript

You will individually write a scientific manuscript (ie short journal
article) of your bacterial genome sequencing work. You will have plenty
of guidance, opportunities for verbal feedback, and be able to submit a
draft version for detailed comment before your final submission. This
format is used across all the project courses, and also by real
scientists publishing their work.

## Course Work

### Labs

You will spend **5 full weeks in labs** (we label these weeks A-E) with
3 in semester 1 (ABC) and 2 (DE) in semester 2. You will be expected to
attend every day during the set hours. If you do not attend fully you
will be letting down the entire group, and will be asked to transfer to
a different module.

We do of course recognise that there will be times when you can't attend
for genuine reasons, and as long as we are made aware then we can come
to some suitable arrangement. Some students will have to step out to
attend lectures on other courses, this will be fine, but please keep us
informed well ahead of time.

Hardy Lab 301 is a specifically designed lab for this type of research.
It will be yours for these 5 weeks. We may also break out to other rooms
for seminars, computing, or whatever works best.

### Study

To really succeed in the course you will need to gain detailed knowledge
as well as lab skills. We will be setting reading, holding informal
journal clubs to discuss papers, and giving seminars to explain the
science. You will need to contribute to these, learning interactively
rather than passively.

## Things for your CV

There are many skills that will look great on your cv. We are happy to
discuss with you how you can best represent your new knowledge and
abilities in this way. Some things that you may be able to include at
the end of the course include:

-   Microbiology techniques. Plating, culturing and gram staining
    > bacteria from clinical isolates

-   Bacterial DNA extraction, construction of DNA sequencing library and
    > running of minION nanopore DNA sequencer

-   Knowledge of the unix command line. Basic command line skills to
    > navigate, inspect and manipulate large data sets

-   Genome assembly and annotation. Basic skills to assemble a bacterial
    > genome from DNA sequence data, and annotate to discover antibiotic
    > resistance genes

-   Reproducible research skills. Understanding of modern reproducible
    > research skills including archiving data and metadata, electronic
    > laboratory records, and open practices.
