# Welcome to Bacterial Genome Sequencing

This is a book to guide you through the Bacterial Genome Sequencing course run for undergraduate students at the University of Hull, UK.

This open resource is licenced CC-BY, meaning that you may use it, adapt it for your own courses, and do mostly as you wish, but we would greatly appreciate you citing the original source.

> Lunt, DH et al (2022) BGS: An undergraduate bacterial genome sequencing course. https:github.com/davelunt/BGS

This course was planned, written, and delivered at the University of Hull, UK, (2022-) by Dave Lunt, Katharine Hubbard, Merideth Freiheit, Graham Sellers, and Mike Winter.

## Who is this book for?

We made this freely available resource primarily for **instructors** trying to develop their own courses. You could however alter it so that it was the student-facing documentation for such a course without many changes.

## Overview of the course

This runs as a 5-week course for final year UK Biology undergraduate students. Our students have already studied some microbiology, and some genetics, but have no knowledge of bioinformatics. The students attend wet lab for 2 full weeks, then computer lab for 2 full weeks, then in the final week carry out their own (mostly wet lab) experiments. We spread these 5 weeks through 2 semesters of study to allow it to be timetabled alongside other courses.

### Week A: Study the bacterial strains

Become familiar with the bacterial strains through Gram staining and microscopy. Finish the week by preparing DNA extractions.

### Week B: nanopore DNA sequencing

Quantify DNA, prepare a sequencing library, load a nanopore sequencer.

### Week C: data QC and genome assembly

Learn to use the unix command line for bioinformatics. Quality control the data, detect contamination, assemble the genome and assess its success.

### Week D: genome annotation and AMR genes

Annotate the genome. Analyse the antimicrobial resistance genes detected.

### Week E: antibiotics

Opportunity for students to test their predictions of antibiotic resistance in the laboratory, to consider employability skills learned, and to problem-solve before writing the final assessed manuscript.

## Notes on this book and other resources

This book is written in markdown and published to html using [Jupyter Book](https://jupyterbook.org). The executable Jupyter notebooks for students are best accessed from a separate [BGS github repository](https://github.com/davelunt/BGS). This repo also contains some other useful files and data. The main data files we generated are too big to easily move around. I hope they are easily downloadable from the International DNA Sequence repositories by the time anyone is reading this, but ultimately you will probably want to use your own genomic data in any case.

![cc-by licence](images/ccby.png)
